File: CPP/14. Longest Common Prefix.cpp
class Solution {
public:
    
    struct Trie{
        int counter = 0;
        Trie* next[26];
        Trie(){
            counter = 0;
            for(int i = 0 ; i< 26; i++){
                next[i] = NULL;
            }
        }
    };
    string longestCommonPrefix(vector<string>& strs) {
        Trie *root = new Trie;
        int len = 0;
        for(int i = 0 ; i< strs.size() ; i++){
            len = max(len, insertTrie(strs[i], root, strs.size()));
        }
        return strs[0].substr(0, len);
    }
    
    int insertTrie(string s, Trie *root, int totalStr){
        Trie *cur = root;
        int maxPreLen = 0; 
        for(int i =  0 ; i< s.size() ; i++){
            int numIthEl = s[i] - 'a';
            
            if(cur -> next[numIthEl] == NULL){
                cur -> next[numIthEl] = new Trie();
            }
            
            cur = cur -> next[numIthEl];
            cur -> counter = cur -> counter + 1;
            if(cur -> counter == totalStr){
                maxPreLen = max(maxPreLen, i+1);
            }
        }
        return maxPreLen;
    }
};

File: CSharp/14. Longest Common Prefix.cs
public class Solution {
    public string LongestCommonPrefix(string[] strs) {
        Trie root = new Trie();
        int result = 0;
        foreach(var x in strs){
            result = Math.Max(result, InsertIntoTrie(x, ref root, strs.Length));
        }
        
        return strs[0].Substring(0,result);
    }
    
    private int InsertIntoTrie(string s, ref Trie root, int totalStr){
        int maxVal = 0;
        
        Trie cur = root;
        for(int i = 0 ; i < s.Length ; i++){
            int numericVal = s[i] - 'a';
            if(cur.next[numericVal] == null){
                cur.next[numericVal] = new Trie();
            }
            cur.next[numericVal].Counter++;
            
            if(cur.next[numericVal].Counter == totalStr){
                maxVal = i+1;
            }
            
            cur = cur.next[numericVal];
        }
        return maxVal;
    }
}


public class Trie
{
    public int Counter { get; set; }
    public Trie[] next;
    public Trie()
    {
        Counter = 0;
        next = new Trie[26];
    }
}